File: services/payment/src/payment/obs.py
"""OpenTelemetry + structured JSON logging for the Payment service.

Every log line carries trace_id/span_id for trace-to-log correlation
(tech-stack §9, C-3..C-8, FR-22, FR-26).
"""

from __future__ import annotations

import json
import logging
import sys
from datetime import datetime, timezone

from opentelemetry import metrics, trace
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator

_propagator = TraceContextTextMapPropagator()
_METRIC_EXPORT_INTERVAL_MS = 15_000


class _JsonTraceFormatter(logging.Formatter):
    """Formats records as JSON, injecting the active trace_id/span_id."""

    def format(self, record: logging.LogRecord) -> str:
        payload: dict[str, object] = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "level": record.levelname,
            "logger": record.name,
            "message": record.getMessage(),
        }
        span = trace.get_current_span()
        ctx = span.get_span_context()
        if ctx.is_valid:
            payload["trace_id"] = format(ctx.trace_id, "032x")
            payload["span_id"] = format(ctx.span_id, "016x")
        if record.exc_info:
            payload["exception"] = self.formatException(record.exc_info)
        return json.dumps(payload)


def setup(service_name: str, otlp_endpoint: str) -> tuple[trace.Tracer, metrics.Meter]:
    """Initialise tracing, metrics, and logging; return the tracer and meter."""
    resource = Resource.create({"service.name": service_name})

    tracer_provider = TracerProvider(resource=resource)
    tracer_provider.add_span_processor(
        BatchSpanProcessor(OTLPSpanExporter(endpoint=otlp_endpoint))
    )
    trace.set_tracer_provider(tracer_provider)

    reader = PeriodicExportingMetricReader(
        OTLPMetricExporter(endpoint=otlp_endpoint),
        export_interval_millis=_METRIC_EXPORT_INTERVAL_MS,
    )
    meter_provider = MeterProvider(resource=resource, metric_readers=[reader])
    metrics.set_meter_provider(meter_provider)

    handler = logging.StreamHandler(sys.stdout)
    handler.setFormatter(_JsonTraceFormatter())
    root = logging.getLogger()
    root.handlers.clear()
    root.addHandler(handler)
    root.setLevel(logging.INFO)

    return trace.get_tracer(service_name), metrics.get_meter(service_name)


def inject_traceparent() -> str:
    """Serialize the active span context into a W3C traceparent string."""
    carrier: dict[str, str] = {}
    _propagator.inject(carrier)
    return carrier.get("traceparent", "")


def context_from_traceparent(traceparent: str):
    """Extract an OTel context from a W3C traceparent header (FR-23)."""
    return _propagator.extract({"traceparent": traceparent}) if traceparent else None

File: services/order/internal/obs/obs.go
// Package obs wires OpenTelemetry (traces + metrics) and structured JSON logging
// with trace_id/span_id correlation (tech-stack §9, C-3..C-8, FR-22, FR-26).
package obs

import (
	"context"
	"log/slog"
	"os"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/propagation"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
	"go.opentelemetry.io/otel/trace"
)

// Providers bundles the OTel providers so main can shut them down cleanly.
type Providers struct {
	Tracer trace.Tracer
	Meter  metric.Meter
	tp     *sdktrace.TracerProvider
	mp     *sdkmetric.MeterProvider
}

// Setup initialises OTLP/gRPC exporters, the W3C propagator, and a slog logger
// whose records include the active trace_id/span_id (FR-26, AC-6).
func Setup(ctx context.Context, serviceName, otlpEndpoint string) (*Providers, *slog.Logger, error) {
	res, err := resource.New(ctx, resource.WithAttributes(semconv.ServiceName(serviceName)))
	if err != nil {
		return nil, nil, err
	}

	traceExp, err := otlptracegrpc.New(ctx,
		otlptracegrpc.WithEndpoint(otlpEndpoint), otlptracegrpc.WithInsecure())
	if err != nil {
		return nil, nil, err
	}
	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(traceExp),
		sdktrace.WithResource(res),
	)
	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.TraceContext{})

	metricExp, err := otlpmetricgrpc.New(ctx,
		otlpmetricgrpc.WithEndpoint(otlpEndpoint), otlpmetricgrpc.WithInsecure())
	if err != nil {
		return nil, nil, err
	}
	mp := sdkmetric.NewMeterProvider(
		sdkmetric.WithResource(res),
		sdkmetric.WithReader(sdkmetric.NewPeriodicReader(metricExp,
			sdkmetric.WithInterval(15*time.Second))),
	)
	otel.SetMeterProvider(mp)

	logger := slog.New(&traceHandler{
		Handler: slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelInfo}),
	}).With("service", serviceName)

	return &Providers{
		Tracer: tp.Tracer(serviceName),
		Meter:  mp.Meter(serviceName),
		tp:     tp,
		mp:     mp,
	}, logger, nil
}

// Shutdown flushes and stops the providers.
func (p *Providers) Shutdown(ctx context.Context) {
	if p.tp != nil {
		_ = p.tp.Shutdown(ctx)
	}
	if p.mp != nil {
		_ = p.mp.Shutdown(ctx)
	}
}

// traceHandler injects trace_id/span_id into every log record for trace-to-log
// correlation in Loki/Tempo (tech-stack §9.2).
type traceHandler struct{ slog.Handler }

func (h *traceHandler) Handle(ctx context.Context, r slog.Record) error {
	if sc := trace.SpanContextFromContext(ctx); sc.IsValid() {
		r.AddAttrs(
			slog.String("trace_id", sc.TraceID().String()),
			slog.String("span_id", sc.SpanID().String()),
		)
	}
	return h.Handler.Handle(ctx, r)
}

func (h *traceHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
	return &traceHandler{Handler: h.Handler.WithAttrs(attrs)}
}

func (h *traceHandler) WithGroup(name string) slog.Handler {
	return &traceHandler{Handler: h.Handler.WithGroup(name)}
}
